server/src/main/java/com/thoughtworks/go/server/service/AuthorizationExtensionCacheService.java

```diff
@@ -0,0 +1,100 @@
+import com.thoughtworks.go.config.SecurityAuthConfig;
+import com.thoughtworks.go.listener.SecurityConfigChangeListener;
+import com.thoughtworks.go.plugin.access.authorization.AuthorizationExtension;
+import com.thoughtworks.go.plugin.domain.authorization.AuthenticationResponse;
+import com.thoughtworks.go.util.SystemEnvironment;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+@Component
+public class AuthorizationExtensionCacheService {
+    private final int CACHE_EXPIRY_IN_MINUTES = SystemEnvironment.getGoServerAuthorizationExtensionCallsCacheTimeoutInSeconds() / 60;
+
+    private final Cache<String, Boolean> isValidUserCache;
+    private final Cache<String, AuthenticationResponse> getUserRolesCache;
+    private final AuthorizationExtension authorizationExtension;
+
+    public AuthorizationExtensionCacheService(GoConfigService goConfigService, AuthorizationExtension authorizationExtension, Ticker ticker) {
+        this.authorizationExtension = authorizationExtension;
+        isValidUserCache = CacheBuilder.newBuilder()
+                .ticker(ticker).expireAfterWrite(CACHE_EXPIRY_IN_MINUTES, TimeUnit.MINUTES).build();
+        getUserRolesCache = CacheBuilder.newBuilder()
+                .ticker(ticker).expireAfterWrite(CACHE_EXPIRY_IN_MINUTES, TimeUnit.MINUTES).build();
+
+        goConfigService.register(this.securityConfigChangeListener());
+    }
+
+    @Autowired
+    public AuthorizationExtensionCacheService(GoConfigService goConfigService, AuthorizationExtension authorizationExtension) {
+        this(goConfigService, authorizationExtension, Ticker.systemTicker());
+    }
+
+    public boolean isValidUser(String pluginId, String username, SecurityAuthConfig authConfig) {
+        String cacheKey = cacheKeyFor(pluginId, username, authConfig);
+        Boolean fromCache = isValidUserCache.getIfPresent(cacheKey);
+
+        if (fromCache == null) {
+            fromCache = authorizationExtension.isValidUser(pluginId, username, authConfig);
+            isValidUserCache.put(cacheKey, fromCache);
+        }
+
+        return fromCache;
+    }
+
+    public AuthenticationResponse getUserRoles(String pluginId, String username, SecurityAuthConfig authConfig, List<PluginRoleConfig> pluginRoleConfigs) {
+        String cacheKey = cacheKeyFor(pluginId, username, authConfig, pluginRoleConfigs);
+        AuthenticationResponse fromCache = getUserRolesCache.getIfPresent(cacheKey);
+
+        if (fromCache == null) {
+            fromCache = authorizationExtension.getUserRoles(pluginId, username, authConfig, pluginRoleConfigs);
+            getUserRolesCache.put(cacheKey, fromCache);
+        }
+
+        return fromCache;
+    }
+
+    private String cacheKeyFor(String pluginId, String username, SecurityAuthConfig authConfig, List<PluginRoleConfig> pluginRoleConfigs) {
```

Comment:
Use `https://github.com/gocd/gocd/blob/c85904bba350615ad5f16b7a39b1d6ba19dad4df/server/src/main/java/com/thoughtworks/go/server/cache/CacheKeyGenerator.java#L27` for cache key generation?